common/src/patch.rs

```diff
@@ -0,0 +1,165 @@
+//! Patch-related functions and types.
+use anyhow::{Error, Result};
+use git2::Oid;
+
+use librad::git::refs::Refs;
+use librad::git::storage::{ReadOnly, ReadOnlyStorage};
+use librad::PeerId;
+
+use crate::{git, person, project};
+use rad_terminal::components as term;
+
+/// A patch can either be open or merged. It's considered merged, as soon as the
+/// merge-base of the default's branch HEAD and the patch commit is the same than the
+/// patch commit itself.
+#[derive(PartialEq, Eq)]
+pub enum State {
+    Open,
+    Merged,
+}
+
+/// A patch is a change set that a user wants the maintainer to merge into a projects default
+/// branch.
+///
+/// A patch is represented by an annotated tag, prefixed with `radicle-patch/`.
+pub struct Metadata {
+    /// ID of a patch. This is the portion of the tag name following the `radicle-patch/` prefix.
+    pub id: String,
+    /// Peer that the patch originated from
+    pub peer: PeerId,
+    /// Message attached to the patch. This is the message of the annotated tag.
+    pub message: Option<String>,
+    /// Head commit that the author wants to merge with this patch.
+    pub commit: Oid,
+    /// The merge base of [`Metadata::commit`] and the head commit of the first maintainer's default
+    /// branch.
+    pub merge_base: Option<Oid>,
+}
+
+impl Metadata {
+    pub fn state(&self) -> State {
+        if self.merge_base.unwrap() == self.commit {
+            return State::Merged;
+        }
+        State::Open
+    }
+
+    pub fn name(&self) -> String {
+        str::replace(&self.id, "radicle-patch/", "")
+    }
+}
+
+/// List patches on the local device. Returns a given peer's patches or this peer's
+/// patches if `peer` is `None`.
+pub fn list<S>(
+    storage: &S,
+    peer: Option<PeerId>,
+    project: &project::Metadata,
+    repo: &git2::Repository,
+) -> Result<Vec<Metadata>, Error>
+where
+    S: AsRef<ReadOnly>,
+{
+    let storage = storage.as_ref();
+    let mut patches: Vec<Metadata> = vec![];
+    let master = repo
+        .resolve_reference_from_short_name(&format!("rad/{}", project.default_branch))?
+        .target()
+        .unwrap();
+
+    if let Ok(refs) = Refs::load(&storage, &project.urn, peer) {
```

Comment:
@cloudhead Is this the correct way to do this? It feels kind of complicated and I'm not 100% sure I covered all cases.

Reply:
This does seem complicated, will have to try to understand what you're trying to do here and get back to you.

Reply:
So basically here we're loading the `signed_refs`, so any ref that was pushed to the monorepo and signed, then we're getting the tag oids, then for each oid we're getting the tag, and getting the merge base between the tag's commit and the repo's default branch. I think this makes sense?

Reply:
Yes, I want to load the `signed_refs` but needed to be able to get a typed tag in the end. And the only way I could figure out was by going through the target object and then peeling a tag out :smirk: 
Is there a better way to do this?